### PR TITLE
[DataGrid] Performance optimization for updating rows={...}

### PR DESCRIPTION
**Pull request by ypresto**
I'm using TextField in 22 columns DataGrid and used Profiler in react-dev-tools to detect hot spot.
This PR reduces re-render for updating rows=[...}

(And bump package.json commit which maybe missing from master branch :)

### CODE REVIEW

File: packages/grid/_modules_/grid/components/Viewport.tsx [hunk 1]
```diff
@@ -42,37 +42,25 @@ export const Viewport: ViewportType = React.forwardRef<HTMLDivElement, {}>(
       if (renderState.renderContext == null) {
         return null;
       }
+      const renderContext = renderState.renderContext as RenderContextProps
```
Comment (ypresto) [inline packages/grid/_modules_/grid/components/Viewport.tsx:45]: Removing Partial<...> here.

File: packages/grid/_modules_/grid/components/RowElements.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  selected,
+  rowHeight,
+}) => {
+  return (
+    <Row
+      className={
+        (renderContext.firstRowIdx + domIndex) % 2 === 0 ? 'Mui-even' : 'Mui-odd'
+      }
+      id={row.id}
+      selected={selected}
```
Comment (ypresto) [inline packages/grid/_modules_/grid/components/RowElements.tsx:34]: Maybe we can remove React.memo() from RowCells by moving selected state lookup to Row element itself, but I'm not familiar with row selection in this library.

**Review by oliviertassinari — COMMENTED** [review]
Thanks for looking into it :).

Could you share a performance benchmark that shows there is a difference? We can't merge any optimizations without a proof they are helping and not hurting (**we never really know until it's measured**). You could use https://codesandbox.io/embed/css-in-js-comparison-forked-5z6e2?codemirror=1&codemirror=1 as a starting point. 

Could you also rebase? There are unrelated changes in the pull request, you are behind.

File: packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx [hunk 3]
```diff
@@ -26,7 +26,7 @@ interface ColumnHeaderItemProps {
   filterItemsCounter?: number;
 }
 
-export const ColumnHeaderItem = ({
+export const ColumnHeaderItem = React.memo(({
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderItem.tsx:29]: An important detail, we can't memoize components that depends on 

```jsx
const apiRef = React.useContext(ApiContext);
```

The logic isn't setup to allow to bypass the memoization when the state change and the component needs to rerender. So, in the current state of things, it's not an option. We have already faced this issue for:

- exposing a composable Toolbar API
- babel plugin that automatically hoists
- and similar cases like this one in the past, e.g. toolbar density menu item

**Comment by oliviertassinari** [comment]
@ypresto I'm closing, for now. Feel free to resume the discussion :)

**Pull request closed without merging**